runtime/src/iree/base/tree.c

```diff
@@ -0,0 +1,549 @@
+  iree_tree_node_t* y = NULL;
+  while (x != &tree->nil) {
+    y = x;
+    if (t->key < x->key) {
+      x = x->left;
+    } else if (t->key > x->key) {
+      x = x->right;
+    } else {
+      return iree_make_status(IREE_STATUS_ALREADY_EXISTS,
+                              "Trying to insert a duplicate key");
+    }
+  }
+  t->parent = y;
+
+  if (!y) {
+    tree->root = t;
+  } else if (t->key < y->key) {
+    y->left = t;
+  } else {
+    y->right = t;
+  }
+
+  if (t->parent == NULL) {
+    t->red = false;
+    return iree_ok_status();
+  }
+
+  if (t->parent == tree->root) {
+    return iree_ok_status();
+  }
+
+  while (t->parent->red) {
+    if (t->parent == t->parent->parent->right) {
+      iree_tree_node_t* uncle = t->parent->parent->left;
+      if (uncle->red) {
+        uncle->red = false;
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        t = t->parent->parent;
+      } else {
+        if (t == t->parent->left) {
+          t = t->parent;
+          iree_tree_rotate_right(tree, t);
+        }
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        iree_tree_rotate_left(tree, t->parent->parent);
+      }
+    } else {
+      iree_tree_node_t* uncle = t->parent->parent->right;
+      if (uncle && uncle->red) {
+        uncle->red = false;
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        t = t->parent->parent;
+      } else {
+        if (t == t->parent->right) {
+          t = t->parent;
+          iree_tree_rotate_left(tree, t);
+        }
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        iree_tree_rotate_right(tree, t->parent->parent);
+      }
+    }
+    if (t == tree->root) {
+      break;
+    }
+  }
+  tree->root->red = false;
+
+  return iree_ok_status();
+}
+
+iree_status_t iree_tree_insert(iree_tree_t* tree, iree_host_size_t key,
+                               iree_tree_node_t** out_data) {
+  iree_tree_node_t* t;
+  iree_status_t status = iree_tree_allocate_node(tree, &t);
+  if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
+    return status;
+  }
+  status = iree_tree_insert_internal(tree, key, t);
+  if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
+    return status;
+  }
+  tree->size++;
+  *out_data = t;
+  return status;
+}
+
+iree_tree_node_t* iree_tree_get(iree_tree_t* tree, iree_host_size_t key) {
+  iree_tree_node_t* node = tree->root;
+  while (node->is_sentinel == false) {
+    if (key == node->key) {
+      return node;
+    } else if (key < node->key) {
+      node = node->left;
+    } else {
+      node = node->right;
+    }
+  }
+  return NULL;
+}
+
+static bool iree_tree_walk_helper(iree_tree_node_t* node,
+                                  iree_tree_walk_type_e walk_type,
+                                  bool (*callback)(iree_tree_node_t*, void*),
+                                  void* user_data) {
+  IREE_ASSERT(walk_type <= IREE_TREE_WALK_POSTORDER);
```

Comment:
IREE_ASSERT_LE (as with IREE_ASSERT_ARGUMENT the intent is to do something better like be able to print out the lhs/rhs like "expected `walk_type` (5) <= `IREE_TREE_WALK_POSTORDER` (4)")